### PR TITLE
compaction: fix transactionindex

### PR DESCRIPTION
**Pull request by chessai**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206062877657852

### CODE REVIEW

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 1]
```diff
@@ -452,6 +457,67 @@ compactionIsIdempotent rdb =
...
+    let makeTx :: Int -> BlockHeader -> IO ChainwebTransaction
+        makeTx nth bh = buildCwCmd testVersion
+          $ set cbSigners [mkEd25519Signer' sender00 [mkGasCap, mkTransferCap "sender00" "sender01" 1.0]]
+          $ setFromHeader bh
+          $ mkCmd (sshow (nth, bh))
+          $ mkExec' "(coin.transfer \"sender00\" \"sender01\" 1.0)"
+
+    supply <- newIORef @Int 0
+    madeTx <- newIORef @Bool False
+    putStrLn "" -- get around tasty formatting
```
Comment (edmundnoble) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:476]: Elaborate please

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 2]
```diff
@@ -452,6 +457,67 @@ compactionIsIdempotent rdb =
...
+              madeTxYet <- readIORef madeTx
+              if madeTxYet
+              then do
+                pure mempty
+              else do
+                n <- atomicModifyIORef' supply $ \a -> (a + 1, a)
+                tx <- makeTx n bh
+                writeIORef madeTx True
+                pure $ V.fromList [tx]
+          }
```
Comment (edmundnoble) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:489]: One day, this sort of pattern will be abstracted over... one day...

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 3]
```diff
@@ -452,6 +457,67 @@ compactionIsIdempotent rdb =
...
+          void $ runBlock cr.pactQueue cr.blockDb second
+          writeIORef madeTx False
+          readIORef supply >>= \s -> when (s `mod` 100 == 0) $ putStrLn $ "ran block " ++ show s
+
+    -- Run enough blocks to where TransactionIndex keeps its full history
+    replicateM_ (int C.transactionIndexKeepDepth) run
```
Comment (edmundnoble) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:495]: Why not make this a configurable parameter to compaction, so that this test is faster?

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs (hunk 1 above)
Comment (chessai) [reply test/Chainweb/Test/Pact/PactSingleChainTest.hs:476]: the test output without this will look like:

```
  testName: first printed line
second printed line
third printed line
```

but with this change, it will look like:
```
  testName:
first printed line
second printed line
third printed line
```

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs (hunk 2 above)
Comment (chessai) [reply test/Chainweb/Test/Pact/PactSingleChainTest.hs:489]: one day...

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs (hunk 3 above)
Comment (chessai) [reply test/Chainweb/Test/Pact/PactSingleChainTest.hs:495]: I didn't think about that. That makes sense.

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs (hunk 1 above)
Comment (chessai) [reply test/Chainweb/Test/Pact/PactSingleChainTest.hs:476]: Maybe there's a better way of doing this. I can't recall. Please suggest if you know of one.

Comment (chessai) [reply test/Chainweb/Test/Pact/PactSingleChainTest.hs]: got rid of the need for this

File: cabal.project [hunk 4]
```diff
@@ -131,6 +138,12 @@ source-repository-package
     tag: 2f67d546ea6608fc6ebe5f2f6976503cbf340442
     --sha256: 0x137akvbh4kr3qagksw74xdj2xz5vjnx1fbr41bb54a0lkcb8mm
```
[the remark this answers is not in the document]
Comment (chessai) [reply cabal.project]: What comment?

[the remark this answers is not in the document]
Comment (chessai) [reply cabal.project]: Nevermind; I removed this change.

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 5]
```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
...
+    replicateM_ keepDepth run
+
+    let db = _sConn cr.sqlEnv
+
+    let getRange :: IO (BlockHeight, BlockHeight)
+        getRange = do
+          r <- Pact.qry db "SELECT MIN(blockheight), MAX(blockheight) FROM TransactionIndex" [] [RInt, RInt]
+          case r of
+            [[SInt mn, SInt mx]] -> pure (int mn, int mx)
+            _ -> assertFailure "getEarliest: invalid query"
```
Comment (chessai) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:505]:
```suggestion
            _ -> assertFailure "getRange: invalid query"
```

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 6]
```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
...
+    let db = _sConn cr.sqlEnv
+
+    let getRange :: IO (BlockHeight, BlockHeight)
+        getRange = do
+          r <- Pact.qry db "SELECT MIN(blockheight), MAX(blockheight) FROM TransactionIndex" [] [RInt, RInt]
+          case r of
+            [[SInt mn, SInt mx]] -> pure (int mn, int mx)
+            _ -> assertFailure "getRange: invalid query"
+
+    let compact = Utils.compact Error [C.NoVacuum, C.TransactionIndexKeepDepth 100] cr.sqlEnv C.Latest
```
Comment (chessai) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:507]:
```suggestion
    let compact = Utils.compact Error [C.NoVacuum, C.TransactionIndexKeepDepth keepDepth] cr.sqlEnv C.Latest
```

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 7]
```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
...
+    do
+      startRange <- getRange
+      compact
+      endRange <- getRange
+      assertEqual "TransactionIndex blockheight range is equal after compaction, when max blockheight is less than or equal to transactionIndexKeepDepth" startRange endRange
+
+    do
+      startRange <- getRange
+      let nMoreBlocks :: Num a => a
+          nMoreBlocks = 100
```
Comment (chessai) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:518]:
```suggestion
          nMoreBlocks = 10
```

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 8]
```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
                   putStrLn ""
       assertFailure "pact state check failed"
 
+compactionTransactionIndex :: ()
+  => RocksDb
+  -> TestTree
+compactionTransactionIndex rdb =
```
Comment (chessai) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:463]: TODO: add another test to attempt to submit a duplicate tx after compaction and see if it is rejected.

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 9]
```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
...
+    let compact = Utils.compact Error [C.NoVacuum, C.TransactionIndexKeepDepth keepDepth] cr.sqlEnv C.Latest
+
+    do
+      startRange <- getRange
+      compact
+      endRange <- getRange
+      assertEqual "TransactionIndex blockheight range is equal after compaction, when max blockheight is less than or equal to transactionIndexKeepDepth" startRange endRange
+
+    do
+      startRange <- getRange
```
Comment (chessai) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:516]: unpack these into eg
```
(preStart, preEnd) <- getRange
...
...
(postStart, postEnd)
```

File: test/Chainweb/Test/Pact/PactSingleChainTest.hs [hunk 10]
```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
...
+      let nMoreBlocks :: Num a => a
+          nMoreBlocks = 10
+      replicateM_ nMoreBlocks run
+      compact
+      endRange <- getRange
+      assertEqual "TransactionIndex blockheight lower range should shift after running more blocks and compacting" (fst startRange + nMoreBlocks) (fst endRange)
```
Comment (chessai) [inline test/Chainweb/Test/Pact/PactSingleChainTest.hs:522]: these top two should be collapsed into one, eg `over both (+ nMoreBlocks) startRange == endRange`

**Comment by edmundnoble** [comment]
Approved pending tx replay test addition.

**Review by edmundnoble — APPROVED** [review]

**Review by thoughtpolice — APPROVED** [review]
Re-approving for merge queue eligibility
